src/app.service.ts

```diff
@@ -0,0 +1,99 @@
+import { Context } from 'koa';
+import { redisGetJson, redisSetJson } from './clients/redis';
+import { LoggerService } from './helpers';
+import { ChannelResult } from './interfaces/channel-result';
+import { CustomerCreditTransferInitiation } from './interfaces/iPain001Transaction';
+import { Channel, NetworkMap } from './interfaces/network-map';
+import { RuleResult } from './interfaces/rule-result';
+import { TypologyResult } from './interfaces/typology-result';
+import apm from 'elastic-apm-node';
+
+export const handleChannels = async (
+  ctx: Context,
+  transaction: CustomerCreditTransferInitiation,
+  networkMap: NetworkMap,
+  ruleResult: RuleResult[],
+  typologyResult: TypologyResult,
+  channelResult: ChannelResult,
+  channel: Channel,
+): Promise<{
+  transactionID: string;
+  message: string;
+}> => {
+  try {
+    const transactionID =
+      transaction.PaymentInformation.CreditTransferTransactionInformation
+        .PaymentIdentification.EndToEndIdentification;
+
+    // Initialize the result message
+    const result = {
+      transactionID: transactionID,
+      message: '',
+    };
+
+    // Check if the channel is completed
+    const hasChannelCompleted = await checkChannelCompletion(
+      transactionID,
+      channel,
+      typologyResult,
+    );
+
+    // If the channel is completed, then save the trasaaction evaluation result
+    if (hasChannelCompleted) {
+      const transactionHistoryQuery = `
+      INSERT {
+        "transactionID": ${JSON.stringify(transactionID)},
+        "transaction": ${JSON.stringify(transaction)},
+        "networkMap": ${JSON.stringify(networkMap)},
+        "ruleResult": ${JSON.stringify(ruleResult)},
+        "typologyResult": ${JSON.stringify(typologyResult)},
+        "channelResult": ${JSON.stringify(channelResult)}
+    } INTO "history"
+    `;
+
+      await ctx.state.arangodb.query(transactionHistoryQuery);
+
+      result.message = 'The transaction evaluation result is saved.';
+    }
+    result.message = 'The transaction evaluation result is not saved.';
+
+    return result;
+  } catch (error) {
+    LoggerService.error(error as string);
+    throw error;
+  }
+};
+
+const checkChannelCompletion = async (
+  transactionID: string,
+  channel: Channel,
+  typologyResult: TypologyResult,
+): Promise<boolean> => {
+  const cacheKey = `${transactionID}_${channel.channel_id}`;
+  const cacheData = await redisGetJson(cacheKey);
```

Comment:
I implemented an interface structure, so that we don't specifically have Redis or ArangoDB specific thing in our logic classes. It's quite simple, but it should make it easier to replace any of the libraries, once we get to that point. Please check what I've done for ADB in https://github.com/ActioFRM/rule-003/pull/19/files#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80